bin/runtime/src/chain_extension/mod.rs

```diff
@@ -1,49 +1,81 @@
         <E::T as SysConfig>::AccountId: UncheckedFrom<<E::T as SysConfig>::Hash> + AsRef<[u8]>,
     {
         match func_id {
-            SNARCOS_CHAIN_EXT => {
-                use pallet_snarcos::{Error, Pallet as Snarcos};
-
-                // The argument-passing-mode doesn't matter for now. All the data to runtime call
-                // are mocked now.
-                let mut env = env.buf_in_buf_out();
-                // After benchmarking is merged and `pallet_snarcos::WeightInfo` is available,
-                // use real weight here.
-                env.charge_weight(41)?;
-
-                match Snarcos::<Runtime>::bare_store_key([0u8; 4], [0u8; 8].to_vec()) {
-                    // In case `DispatchResultWithPostInfo` was returned (or some simpler
-                    // equivalent for `bare_store_key`), we could adjust weight. However, for the
-                    // storing key action it doesn't make sense.
-                    Ok(_) => {
-                        // Return status code for success.
-                        Ok(RetVal::Converging(0))
-                    }
-                    Err(Error::<Runtime>::VerificationKeyTooLong) => Ok(RetVal::Converging(1)),
-                    Err(Error::<Runtime>::IdentifierAlreadyInUse) => Ok(RetVal::Converging(2)),
-                    // Unknown error.
-                    _ => Ok(RetVal::Converging(3)),
-                }
-            }
+            SNARCOS_STORE_KEY_FUNC_ID => Self::snarcos_store_key(env),
             _ => {
                 error!("Called an unregistered `func_id`: {}", func_id);
                 Err(DispatchError::Other("Unimplemented func_id"))
             }
         }
     }
 }
+
+pub type ByteCount = u32;
+
+impl SnarcosChainExtension {
+    fn store_key_weight(key_length: ByteCount) -> Weight {
+        <<Runtime as Config>::WeightInfo as WeightInfo>::store_key(key_length)
+    }
+
+    fn snarcos_store_key<E: Ext>(env: Environment<E, InitState>) -> Result<RetVal, DispatchError>
+    where
+        <E::T as SysConfig>::AccountId: UncheckedFrom<<E::T as SysConfig>::Hash> + AsRef<[u8]>,
+    {
+        // We need to read input as plain bytes (encoded args).
+        let mut env = env.buf_in_buf_out();
+
+        // Check if it makes sense to start reading and decoding data.
+        let key_length = env
+            .in_len()
+            .saturating_sub(size_of::<VerificationKeyIdentifier>() as ByteCount);
+        if key_length > MaximumVerificationKeyLength::get() {
+            return Ok(RetVal::Converging(SNARCOS_STORE_KEY_TOO_LONG_KEY));
+        }
+
+        // We can decode identifier without charging yet - it is cheap and doesn't interact with
+        // any storage.
+        let identifier = env.read_as::<VerificationKeyIdentifier>()?;
+
+        // Now we charge - even if decoding fails and we shouldn't touch storage, we have to incur
+        // fee for reading memory.
+        env.charge_weight(Self::store_key_weight(key_length))?;
```

Comment:
Reading is probably cheaper than storing so we should revisit in the future?

Reply:
well yes, but it will be quite weird to benchmark; to be rethought